This is an example command taken from `gs://broad-pharma5-ukbb-outputs/ukb-freeze-4-fe.all/data/CRAM/CRAM18/UKB_5700695_236231607.g.vcf.gz`.
The GATK version in this command and command line args were used to create the UKBB GATK Docker image and to  write the HaplotypeCaller command in `run_haplotype_caller.py`.

```
##GATKCommandLine=<ID=HaplotypeCaller,CommandLine="HaplotypeCaller  --emit-ref-confidence GVCF --gvcf-gq-bands 10 --gvcf-gq-bands 20 --gvcf-gq-bands 30 --gvcf-gq-bands 40 --gvcf-gq-bands 50 --gvcf-gq-bands 60 --gvcf-gq-bands 70 --gvcf-gq-bands 80 --gvcf-gq-bands 90 --use-new-qual-calculator true --contamination-fraction-to-filter 0.0069722000000000004 --output UKB_5700695_236231607.g.vcf.gz --intervals /cromwell_root/broad-pharma5-execution5/WhiteAlbumExomeReprocessing/b339c755-4a66-4a55-b2dc-6edd9fe9e508/call-ExomeReprocessing/ExomeReprocessing/cad88cf2-116c-4bb0-8e81-59ed5d55509f/call-ExomeGermlineSingleSample/ExomeGermlineSingleSample/bed240fb-6f43-441e-b57c-9f2012c1f6d7/call-BamToGvcf/BamToGvcf/b121bce5-9833-4dbe-afd5-a8f9e8b4ee72/call-ScatterIntervalList/glob-cb4648beeaff920acb03de7603c06f98/41scattered.interval_list --input gs://broad-pharma5-execution5/WhiteAlbumExomeReprocessing/b339c755-4a66-4a55-b2dc-6edd9fe9e508/call-ExomeReprocessing/ExomeReprocessing/cad88cf2-116c-4bb0-8e81-59ed5d55509f/call-ExomeGermlineSingleSample/ExomeGermlineSingleSample/bed240fb-6f43-441e-b57c-9f2012c1f6d7/call-UnmappedBamToAlignedBam/UnmappedBamToAlignedBam/1528268a-6ec0-4661-b23a-14454b34e5e3/call-GatherBamFiles/UKB_5700695_236231607.bam --reference /cromwell_root/broad-references/hg38/v0/Homo_sapiens_assembly38.fasta --annotation-group StandardAnnotation --annotation-group StandardHCAnnotation --annotation-group AS_StandardAnnotation  --indel-size-to-eliminate-in-ref-model 10 --use-alleles-trigger false --disable-optimizations false --just-determine-active-regions false --dont-genotype false --max-mnp-distance 0 --dont-trim-active-regions false --max-disc-ar-extension 25 --max-gga-ar-extension 300 --padding-around-indels 150 --padding-around-snps 20 --kmer-size 10 --kmer-size 25 --dont-increase-kmer-sizes-for-cycles false --allow-non-unique-kmers-in-ref false --num-pruning-samples 1 --recover-dangling-heads false --do-not-recover-dangling-branches false --min-dangling-branch-length 4 --consensus false --max-num-haplotypes-in-population 128 --error-correct-kmers false --min-pruning 2 --debug-graph-transformations false --kmer-length-for-read-error-correction 25 --min-observations-for-kmer-to-be-solid 20 --likelihood-calculation-engine PairHMM --base-quality-score-threshold 18 --pair-hmm-gap-continuation-penalty 10 --pair-hmm-implementation FASTEST_AVAILABLE --pcr-indel-model CONSERVATIVE --phred-scaled-global-read-mismapping-rate 45 --native-pair-hmm-threads 4 --native-pair-hmm-use-double-precision false --debug false --use-filtered-reads-for-annotations false --bam-writer-type CALLED_HAPLOTYPES --dont-use-soft-clipped-bases false --capture-assembly-failure-bam false --error-correct-reads false --do-not-run-physical-phasing false --min-base-quality-score 10 --smith-waterman JAVA --annotate-with-num-discovered-alleles false --heterozygosity 0.001 --indel-heterozygosity 1.25E-4 --heterozygosity-stdev 0.01 --standard-min-confidence-threshold-for-calling 10.0 --max-alternate-alleles 6 --max-genotype-count 1024 --sample-ploidy 2 --num-reference-samples-if-no-call 0 --genotyping-mode DISCOVERY --genotype-filtered-alleles false --output-mode EMIT_VARIANTS_ONLY --all-site-pls false --min-assembly-region-size 50 --max-assembly-region-size 300 --assembly-region-padding 100 --max-reads-per-alignment-start 50 --active-probability-threshold 0.002 --max-prob-propagation-distance 50 --interval-set-rule UNION --interval-padding 0 --interval-exclusion-padding 0 --interval-merging-rule ALL --read-validation-stringency SILENT --seconds-between-progress-updates 10.0 --disable-sequence-dictionary-validation false --create-output-bam-index true --create-output-bam-md5 false --create-output-variant-index true --create-output-variant-md5 false --lenient false --add-output-sam-program-record true --add-output-vcf-command-line true --cloud-prefetch-buffer 40 --cloud-index-prefetch-buffer -1 --disable-bam-index-caching false --sites-only-vcf-output false --help false --version false --showHidden false --verbosity INFO --QUIET false --use-jdk-deflater false --use-jdk-inflater false --gcs-max-retries 20 --gcs-project-for-requester-pays  --disable-tool-default-read-filters false --minimum-mapping-quality 20 --disable-tool-default-annotations false --enable-all-annotations false",Version=4.0.10.1,Date="February 2, 2019 1:07:25 PM UTC">
```
